File: aws/attack-scripts/iot-attack-emulator.ts
#!/usr/bin/env node
/**
 * IoT Attack Emulation Script for Network Threat Detection Testbed
 * TypeScript implementation optimized for network operations
 */

import * as mqtt from 'mqtt';
import * as AWS from 'aws-sdk';
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { v4 as uuidv4 } from 'uuid';

interface DeviceConfig {
  deviceId: string;
  deviceType: string;
  baseTemp: number;
  tempVariance: number;
  normalInterval: number;
  isCompromised: boolean;
  attackType?: AttackType;
}

interface AttackPayload {
  timestamp: number;
  deviceId: string;
  deviceType: string;
  temperature: number;
  status: string;
  sequenceNumber: number;
  firmwareVersion: string;
  batteryLevel?: number;
  attackSignature?: string;
  [key: string]: any;
}

type AttackType = 'ddos_flood' | 'data_exfiltration' | 'c2_beacon' | 'scanning' | 'backdoor' | 'injection';

type DeviceType = 'thermostat' | 'sensor' | 'camera' | 'tracker' | 'gateway' | 'actuator';

class IoTDevice {
  private config: DeviceConfig;
  private mqttClient: mqtt.MqttClient;
  private topic: string;
  private running: boolean = false;
  private attackTimer?: NodeJS.Timeout;
  private normalTimer?: NodeJS.Timeout;
  private sequenceNumber: number = 0;

  constructor(
    deviceId: string,
    deviceType: DeviceType,
    mqttClient: mqtt.MqttClient,
    topicPrefix: string = 'device'
  ) {
    this.config = {
      deviceId,
      deviceType,
      baseTemp: 20 + Math.random() * 6, // 20-26°C
      tempVariance: 0.5 + Math.random() * 1.5, // 0.5-2°C
      normalInterval: 30000 + Math.random() * 90000, // 30-120 seconds
      isCompromised: false
    };
    
    this.mqttClient = mqttClient;
    this.topic = `${topicPrefix}/${deviceId}/data`;
    
    console.log(`Created device ${deviceId} of type ${deviceType}`);
  }

  private generateBenignPayload(): AttackPayload {
    const temp = this.config.baseTemp + (Math.random() - 0.5) * 2 * this.config.tempVariance;
    const statuses = ['NORMAL', 'IDLE', 'ACTIVE'];
    const status = statuses[Math.floor(Math.random() * statuses.length)];
    
    const payload: AttackPayload = {
      timestamp: Date.now(),
      deviceId: this.config.deviceId,
      deviceType: this.config.deviceType,
      temperature: Math.round(temp * 100) / 100,
      status,
      sequenceNumber: ++this.sequenceNumber,
      firmwareVersion: '1.2.3'
    };

    // Add battery level for mobile devices
    if (['sensor', 'tracker'].includes(this.config.deviceType)) {
      payload.batteryLevel = Math.round((85 + Math.random() * 15) * 10) / 10;
    }

    return payload;
  }

  private generateMaliciousPayload(attackType: AttackType): AttackPayload {
    const basePayload = this.generateBenignPayload();
    
    switch (attackType) {
      case 'ddos_flood':
        return {
          ...basePayload,
          status: 'ATTACK_DDOS',
          attackSignature: 'ddos_flood',
          packetSize: Math.floor(Math.random() * 1024) + 64, // 64-1088 bytes
          targetEndpoint: '192.168.1.1',
          floodRate: Math.floor(Math.random() * 1000) + 100 // packets/sec
        };

      case 'data_exfiltration':
        const sensitiveData = {
          userCredentials: 'admin:password123',
          networkConfig: '192.168.1.0/24',
          deviceKeys: uuidv4(),
          wifiPasswords: ['HomeWiFi123', 'GuestNetwork456'],
          locationData: { lat: 40.7128, lng: -74.0060 }
        };
        
        return {
          ...basePayload,
          status: 'ATTACK_EXFIL',
          attackSignature: 'data_exfiltration',
          exfiltratedData: 'x'.repeat(1024 + Math.floor(Math.random() * 4096)), // 1-5KB
          sensitiveInfo: sensitiveData,
          exfilMethod: Math.random() > 0.5 ? 'dns_tunneling' : 'http_post',
          encryptionUsed: Math.random() > 0.3
        };

      case 'c2_beacon':
        const commands = ['heartbeat', 'update', 'execute', 'download', 'upload'];
        return {
          ...basePayload,
          status: 'ATTACK_C2',
          attackSignature: 'c2_beacon',
          temperature: -999, // Unusual value
          c2Command: commands[Math.floor(Math.random() * commands.length)],
          c2Server: '192.168.99.100',
          beaconInterval: Math.floor(Math.random() * 300) + 60, // 1-5 minutes
          payload: crypto.randomBytes(16).toString('hex')
        };

      case 'scanning':
        const commonPorts = [21, 22, 23, 25, 53, 80, 110, 143, 443, 993, 995, 1433, 3389];
        const scannedPorts = Array.from({ length: 10 }, () => 
          Math.floor(Math.random() * 65535) + 1
        );
        
        return {
          ...basePayload,
          status: 'ATTACK_SCAN',
          attackSignature: 'network_scanning',
          scannedPorts: [...commonPorts, ...scannedPorts],
          targetNetwork: '192.168.1.0/24',
          scanType: Math.random() > 0.5 ? 'tcp_syn' : 'tcp_connect',
          portResponses: scannedPorts.map(port => ({
            port,
            status: Math.random() > 0.8 ? 'open' : 'closed'
          }))
        };

      case 'backdoor':
        const backdoorCommands = ['shell', 'file_access', 'privilege_escalation', 'persistence'];
        return {
          ...basePayload,
          status: 'ATTACK_BACKDOOR',
          attackSignature: 'backdoor_access',
          unauthorizedAccess: true,
          backdoorCommand: backdoorCommands[Math.floor(Math.random() * backdoorCommands.length)],
          executionTime: Math.floor(Math.random() * 10000) + 1000,
          privilegeLevel: Math.random() > 0.5 ? 'admin' : 'user',
          persistenceMechanism: Math.random() > 0.5 ? 'service' : 'registry'
        };

      case 'injection':
        const injectionPayloads = [
          "'; DROP TABLE users; --",
          '<script>alert("XSS")</script>',
          '$(rm -rf /)',
          '../../../etc/passwd',
          'admin\' OR \'1\'=\'1'
        ];
        
        return {
          ...basePayload,
          status: 'ATTACK_INJECTION',
          attackSignature: 'code_injection',
          injectedCode: injectionPayloads[Math.floor(Math.random() * injectionPayloads.length)],
          injectionType: ['sql', 'xss', 'command', 'path_traversal', 'ldap'][Math.floor(Math.random() * 5)],
          targetField: ['username', 'password', 'search', 'filename'][Math.floor(Math.random() * 4)],
          bypassAttempt: Math.random() > 0.3
        };

      default:
        return basePayload;
    }
  }

  private async publishMessage(payload: AttackPayload): Promise<void> {
    return new Promise((resolve, reject) => {
      const message = JSON.stringify(payload, null, 2);
      
      this.mqttClient.publish(this.topic, message, { qos: 1 }, (error) => {
        if (error) {
          console.error(`Failed to publish message from ${this.config.deviceId}:`, error);
          reject(error);
        } else {
          console.log(`[${this.config.deviceId}] Published: ${payload.status}`);
          resolve();
        }
      });
    });
  }

  public startBenignBehavior(): void {
    if (this.running) return;
    
    this.running = true;
    console.log(`Starting benign behavior for device ${this.config.deviceId}`);
    
    const benignLoop = async () => {
      if (!this.running || this.config.isCompromised) return;
      
      try {
        const payload = this.generateBenignPayload();
        await this.publishMessage(payload);
      } catch (error) {
        console.error(`Error in benign loop for ${this.config.deviceId}:`, error);
      }
      
      if (this.running && !this.config.isCompromised) {
        this.normalTimer = setTimeout(benignLoop, this.config.normalInterval);
      }
    };
    
    benignLoop();
  }

  public async startAttackBehavior(attackType: AttackType, duration: number = 60000): Promise<void> {
    console.warn(`Device ${this.config.deviceId} compromised - starting ${attackType} attack`);
    
    this.config.isCompromised = true;
    this.config.attackType = attackType;
    
    // Clear normal timer
    if (this.normalTimer) {
      clearTimeout(this.normalTimer);
      this.normalTimer = undefined;
    }
    
    const attackLoop = async () => {
      if (!this.running || !this.config.isCompromised) return;
      
      try {
        const payload = this.generateMaliciousPayload(attackType);
        await this.publishMessage(payload);
      } catch (error) {
        console.error(`Error in attack loop for ${this.config.deviceId}:`, error);
      }
      
      if (this.running && this.config.isCompromised) {
        // Different attack frequencies
        let nextInterval: number;
        switch (attackType) {
          case 'ddos_flood':
            nextInterval = 100 + Math.random() * 400; // 100-500ms
            break;
          case 'c2_beacon':
            nextInterval = 10000 + Math.random() * 20000; // 10-30s
            break;
          case 'scanning':
            nextInterval = 1000 + Math.random() * 2000; // 1-3s
            break;
          default:
            nextInterval = 5000 + Math.random() * 10000; // 5-15s
        }
        
        this.attackTimer = setTimeout(attackLoop, nextInterval);
      }
    };
    
    // Start attack immediately
    attackLoop();
    
    // Schedule return to normal behavior
    setTimeout(() => {
      this.stopAttackBehavior();
    }, duration);
  }

  public stopAttackBehavior(): void {
    if (!this.config.isCompromised) return;
    
    console.log(`Device ${this.config.deviceId} attack behavior ended - returning to normal`);
    
    this.config.isCompromised = false;
    this.config.attackType = undefined;
    
    // Clear attack timer
    if (this.attackTimer) {
      clearTimeout(this.attackTimer);
      this.attackTimer = undefined;
    }
    
    // Resume normal behavior
    if (this.running) {
      this.startBenignBehavior();
    }
  }

  public stop(): void {
    console.log(`Stopping device ${this.config.deviceId}`);
    
    this.running = false;
    
    if (this.normalTimer) {
      clearTimeout(this.normalTimer);
      this.normalTimer = undefined;
    }
    
    if (this.attackTimer) {
      clearTimeout(this.attackTimer);
      this.attackTimer = undefined;
    }
  }
}

class AttackEmulator {
  private iotEndpoint: string;
  private numDevices: number;
  private devices: IoTDevice[] = [];
  private mqttClient?: mqtt.MqttClient;
  private running: boolean = false;
  
  private readonly attackScenarios: AttackType[] = [
    'ddos_flood',
    'data_exfiltration',
    'c2_beacon',
    'scanning',
    'backdoor',
    'injection'
  ];
  
  private readonly deviceTypes: DeviceType[] = [
    'thermostat',
    'sensor',
    'camera',
    'tracker',
    'gateway',
    'actuator'
  ];

  constructor(iotEndpoint: string, numDevices: number = 10) {
    this.iotEndpoint = iotEndpoint;
    this.numDevices = numDevices;
  }

  private setupMqttClient(): Promise<void> {
    return new Promise((resolve, reject) => {
      console.log(`Connecting to AWS IoT Core: ${this.iotEndpoint}`);
      
      // Setup MQTT client with AWS IoT Core
      const clientOptions: mqtt.IClientOptions = {
        port: 8883,
        protocol: 'mqtts',
        rejectUnauthorized: false,
        reconnectPeriod: 5000,
        connectTimeout: 30000,
        keepalive: 60
      };

      this.mqttClient = mqtt.connect(`mqtts://${this.iotEndpoint}`, clientOptions);
      
      this.mqttClient.on('connect', () => {
        console.log('Connected to AWS IoT Core successfully');
        resolve();
      });
      
      this.mqttClient.on('error', (error) => {
        console.error('MQTT connection error:', error);
        reject(error);
      });
      
      this.mqttClient.on('disconnect', () => {
        console.warn('Disconnected from AWS IoT Core');
      });
      
      this.mqttClient.on('reconnect', () => {
        console.log('Reconnecting to AWS IoT Core...');
      });
    });
  }

  private createDevices(): void {
    console.log(`Creating ${this.numDevices} simulated IoT devices...`);
    
    for (let i = 0; i < this.numDevices; i++) {
      const deviceId = `device_${i.toString().padStart(3, '0')}`;
      const deviceType = this.deviceTypes[Math.floor(Math.random() * this.deviceTypes.length)];
      
      const device = new IoTDevice(deviceId, deviceType, this.mqttClient!);
      this.devices.push(device);
    }
    
    console.log(`Created ${this.devices.length} devices`);
  }

  private startBenignSimulation(): void {
    console.log('Starting benign traffic simulation...');
    
    this.devices.forEach(device => {
      device.startBenignBehavior();
    });
    
    console.log(`Started benign simulation for ${this.devices.length} devices`);
  }

  private async launchAttackScenario(
    attackType: AttackType,
    numCompromised?: number,
    duration: number = 60000
  ): Promise<IoTDevice[]> {
    if (!numCompromised) {
      numCompromised = Math.floor(Math.random() * Math.max(1, this.devices.length / 3)) + 1;
    }
    
    // Select random devices to compromise
    const shuffled = [...this.devices].sort(() => 0.5 - Math.random());
    const compromisedDevices = shuffled.slice(0, numCompromised);
    
    console.warn(`Launching ${attackType} attack on ${numCompromised} devices for ${duration}ms`);
    
    // Start attack behavior on compromised devices
    const attackPromises = compromisedDevices.map(device => 
      device.startAttackBehavior(attackType, duration)
    );
    
    await Promise.all(attackPromises);
    
    return compromisedDevices;
  }

  public async runContinuousSimulation(attackFrequency: number = 300000): Promise<void> {
    console.log('Starting continuous attack simulation...');
    this.running = true;
    
    // Start benign simulation
    this.startBenignSimulation();
    
    // Periodic attack simulation
    while (this.running) {
      try {
        // Random interval between attacks
        const sleepTime = attackFrequency * 0.5 + Math.random() * attackFrequency;
        console.log(`Waiting ${Math.round(sleepTime / 1000)}s until next attack scenario...`);
        
        await new Promise(resolve => setTimeout(resolve, sleepTime));
        
        if (!this.running) break;
        
        // Random attack scenario
        const attackType = this.attackScenarios[Math.floor(Math.random() * this.attackScenarios.length)];
        const numCompromised = Math.floor(Math.random() * Math.max(1, this.devices.length / 4)) + 1;
        const duration = 30000 + Math.random() * 90000; // 30-120 seconds
        
        await this.launchAttackScenario(attackType, numCompromised, duration);
        
      } catch (error) {
        console.error('Error in continuous simulation:', error);
        await new Promise(resolve => setTimeout(resolve, 30000));
      }
    }
  }

  public async runSingleAttackScenario(
    attackType: AttackType,
    duration: number = 60000
  ): Promise<void> {
    console.log(`Running single attack scenario: ${attackType}`);
    
    // Start benign simulation
    this.startBenignSimulation();
    
    // Allow benign traffic to establish
    await new Promise(resolve => setTimeout(resolve, 10000));
    
    // Launch attack
    await this.launchAttackScenario(attackType, undefined, duration);
    
    // Keep running for duration + buffer
    await new Promise(resolve => setTimeout(resolve, duration + 30000));
  }

  public async initialize(): Promise<void> {
    try {
      await this.setupMqttClient();
      this.createDevices();
      console.log('Attack emulator initialized successfully');
    } catch (error) {
      console.error('Failed to initialize attack emulator:', error);
      throw error;
    }
  }

  public stop(): void {
    console.log('Stopping attack emulation...');
    this.running = false;
    
    // Stop all devices
    this.devices.forEach(device => device.stop());
    
    // Disconnect MQTT client
    if (this.mqttClient) {
      this.mqttClient.end();
    }
    
    console.log('Attack emulation stopped');
  }
}

// CLI Interface
async function main() {
  const args = process.argv.slice(2);
  
  // Parse command line arguments
  const endpoint = getArg(args, '--endpoint');
  const numDevices = parseInt(getArg(args, '--devices', '10'));
  const attackFrequency = parseInt(getArg(args, '--attack-frequency', '300')) * 1000;
  const mode = getArg(args, '--mode', 'continuous');
  const attackType = getArg(args, '--attack-type') as AttackType;
  const duration = parseInt(getArg(args, '--duration', '60')) * 1000;
  
  if (!endpoint) {
    console.error('Error: --endpoint is required');
    process.exit(1);
  }
  
  const emulator = new AttackEmulator(endpoint, numDevices);
  
  // Handle graceful shutdown
  process.on('SIGINT', () => {
    console.log('\nReceived interrupt signal, stopping simulation...');
    emulator.stop();
    process.exit(0);
  });
  
  process.on('SIGTERM', () => {
    console.log('\nReceived termination signal, stopping simulation...');
    emulator.stop();
    process.exit(0);
  });
  
  try {
    await emulator.initialize();
    
    if (mode === 'continuous') {
      await emulator.runContinuousSimulation(attackFrequency);
    } else {
      const selectedAttackType = attackType || 
        ['ddos_flood', 'data_exfiltration', 'c2_beacon', 'scanning', 'backdoor', 'injection'][
          Math.floor(Math.random() * 6)
        ] as AttackType;
      
      await emulator.runSingleAttackScenario(selectedAttackType, duration);
    }
    
  } catch (error) {
    console.error('Error in main execution:', error);
    process.exit(1);
  }
}

function getArg(args: string[], name: string, defaultValue?: string): string {
  const index = args.indexOf(name);
  if (index !== -1 && index + 1 < args.length) {
    return args[index + 1];
  }
  return defaultValue || '';
}

// Run if called directly
if (require.main === module) {
  main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

export { AttackEmulator, IoTDevice, AttackType, DeviceType };

File: aws/attack-scripts/iot_attack_emulator.py
#!/usr/bin/env python3
"""
IoT Attack Emulation Script for Network Threat Detection Testbed
Generates both benign and malicious IoT traffic for model validation.
"""

import json
import random
import time
import threading
import logging
import argparse
import boto3
from datetime import datetime
from typing import Dict, List, Optional
import paho.mqtt.client as mqtt
import ssl
import uuid

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class IoTDevice:
    """Represents a simulated IoT device that can generate benign or malicious traffic."""
    
    def __init__(self, device_id: str, device_type: str, 
                 mqtt_client: mqtt.Client, topic_prefix: str = "device"):
        self.device_id = device_id
        self.device_type = device_type
        self.mqtt_client = mqtt_client
        self.topic = f"{topic_prefix}/{device_id}/data"
        self.is_compromised = False
        self.attack_type = None
        self.running = True
        
        # Device-specific parameters
        self.base_temp = random.uniform(20.0, 26.0)
        self.temp_variance = random.uniform(0.5, 2.0)
        self.normal_interval = random.uniform(30, 120)  # seconds
        
    def generate_benign_payload(self) -> Dict:
        """Generate normal IoT device telemetry data."""
        # Temperature with realistic variations
        temp = self.base_temp + random.uniform(-self.temp_variance, self.temp_variance)
        
        # Realistic device status
        status = random.choice(['NORMAL', 'IDLE', 'ACTIVE'])
        
        # Battery level (if applicable)
        battery = random.uniform(85, 100) if self.device_type in ['sensor', 'tracker'] else None
        
        payload = {
            'timestamp': int(time.time()),
            'device_id': self.device_id,
            'device_type': self.device_type,
            'temperature': round(temp, 2),
            'status': status,
            'sequence_number': random.randint(1, 10000),
            'firmware_version': '1.2.3'
        }
        
        if battery:
            payload['battery_level'] = round(battery, 1)
            
        return payload
    
    def generate_malicious_payload(self, attack_type: str) -> Dict:
        """Generate malicious traffic based on attack type."""
        base_payload = self.generate_benign_payload()
        
        if attack_type == 'ddos_flood':
            # DDoS: High frequency, small payloads
            base_payload['status'] = 'ATTACK_DDOS'
            base_payload['attack_signature'] = 'ddos_flood'
            
        elif attack_type == 'data_exfiltration':
            # Data exfiltration: Large payloads with sensitive data
            base_payload['status'] = 'ATTACK_EXFIL'
            base_payload['attack_signature'] = 'data_exfiltration'
            # Simulate exfiltrated data
            base_payload['exfiltrated_data'] = 'x' * random.randint(1024, 5120)  # 1-5KB
            base_payload['sensitive_info'] = {
                'user_credentials': 'admin:password123',
                'network_config': '192.168.1.0/24',
                'device_keys': str(uuid.uuid4())
            }
            
        elif attack_type == 'c2_beacon':
            # C2 beacons: Small, periodic, with unusual values
            base_payload['status'] = 'ATTACK_C2'
            base_payload['attack_signature'] = 'c2_beacon'
            base_payload['temperature'] = -999  # Unusual value
            base_payload['c2_command'] = random.choice(['heartbeat', 'update', 'execute'])
            base_payload['c2_server'] = '192.168.99.100'
            
        elif attack_type == 'scanning':
            # Network scanning: Port scanning patterns
            base_payload['status'] = 'ATTACK_SCAN'
            base_payload['attack_signature'] = 'network_scanning'
            base_payload['scanned_ports'] = random.sample(range(1, 65535), 10)
            base_payload['target_network'] = '192.168.1.0/24'
            
        elif attack_type == 'backdoor':
            # Backdoor: Unauthorized access patterns
            base_payload['status'] = 'ATTACK_BACKDOOR'
            base_payload['attack_signature'] = 'backdoor_access'
            base_payload['unauthorized_access'] = True
            base_payload['backdoor_command'] = random.choice(['shell', 'file_access', 'privilege_escalation'])
            
        elif attack_type == 'injection':
            # Injection attacks: Malicious code injection
            base_payload['status'] = 'ATTACK_INJECTION'
            base_payload['attack_signature'] = 'code_injection'
            base_payload['injected_code'] = "'; DROP TABLE users; --"
            base_payload['injection_type'] = random.choice(['sql', 'command', 'script'])
            
        return base_payload
    
    def publish_message(self, payload: Dict) -> None:
        """Publish message to MQTT broker."""
        try:
            message = json.dumps(payload)
            result = self.mqtt_client.publish(self.topic, message, qos=1)
            
            if result.rc == mqtt.MQTT_ERR_SUCCESS:
                logger.debug(f"Published message from {self.device_id}: {payload['status']}")
            else:
                logger.error(f"Failed to publish message from {self.device_id}: {result.rc}")
                
        except Exception as e:
            logger.error(f"Error publishing message from {self.device_id}: {e}")
    
    def start_benign_behavior(self) -> None:
        """Start normal device behavior in a separate thread."""
        def benign_loop():
            while self.running:
                if not self.is_compromised:
                    payload = self.generate_benign_payload()
                    self.publish_message(payload)
                    time.sleep(self.normal_interval)
                else:
                    time.sleep(1)  # Short sleep when compromised
        
        thread = threading.Thread(target=benign_loop, daemon=True)
        thread.start()
        logger.info(f"Started benign behavior for device {self.device_id}")
    
    def start_attack_behavior(self, attack_type: str, duration: int = 60) -> None:
        """Start attack behavior for specified duration."""
        def attack_loop():
            self.is_compromised = True
            self.attack_type = attack_type
            start_time = time.time()
            
            logger.warning(f"Device {self.device_id} compromised - starting {attack_type} attack")
            
            while self.running and (time.time() - start_time) < duration:
                payload = self.generate_malicious_payload(attack_type)
                self.publish_message(payload)
                
                # Different attack patterns have different frequencies
                if attack_type == 'ddos_flood':
                    time.sleep(random.uniform(0.1, 0.5))  # High frequency
                elif attack_type == 'c2_beacon':
                    time.sleep(random.uniform(10, 30))  # Periodic beacons
                elif attack_type == 'scanning':
                    time.sleep(random.uniform(1, 3))  # Moderate frequency
                else:
                    time.sleep(random.uniform(5, 15))  # Low frequency
            
            # Return to normal behavior
            self.is_compromised = False
            self.attack_type = None
            logger.info(f"Device {self.device_id} attack behavior ended - returning to normal")
        
        thread = threading.Thread(target=attack_loop, daemon=True)
        thread.start()
    
    def stop(self) -> None:
        """Stop device simulation."""
        self.running = False
        logger.info(f"Stopped device {self.device_id}")


class AttackEmulator:
    """Main attack emulation orchestrator."""
    
    def __init__(self, iot_endpoint: str, num_devices: int = 10):
        self.iot_endpoint = iot_endpoint
        self.num_devices = num_devices
        self.devices: List[IoTDevice] = []
        self.mqtt_client = None
        self.running = False
        
        # Attack scenario configurations
        self.attack_scenarios = [
            'ddos_flood',
            'data_exfiltration',
            'c2_beacon',
            'scanning',
            'backdoor',
            'injection'
        ]
        
        # Device types for realistic simulation
        self.device_types = [
            'thermostat',
            'sensor',
            'camera',
            'tracker',
            'gateway',
            'actuator'
        ]
    
    def setup_mqtt_client(self) -> None:
        """Setup MQTT client with AWS IoT Core."""
        self.mqtt_client = mqtt.Client()
        
        # Setup TLS
        context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
        context.check_hostname = False
        context.verify_mode = ssl.CERT_NONE
        
        self.mqtt_client.tls_set_context(context)
        
        # Callbacks
        self.mqtt_client.on_connect = self.on_connect
        self.mqtt_client.on_disconnect = self.on_disconnect
        self.mqtt_client.on_publish = self.on_publish
        
        try:
            logger.info(f"Connecting to AWS IoT Core: {self.iot_endpoint}")
            self.mqtt_client.connect(self.iot_endpoint, 8883, 60)
            self.mqtt_client.loop_start()
            
        except Exception as e:
            logger.error(f"Failed to connect to AWS IoT Core: {e}")
            raise
    
    def on_connect(self, client, userdata, flags, rc):
        """MQTT connection callback."""
        if rc == 0:
            logger.info("Connected to AWS IoT Core successfully")
        else:
            logger.error(f"Failed to connect to AWS IoT Core: {rc}")
    
    def on_disconnect(self, client, userdata, rc):
        """MQTT disconnection callback."""
        logger.warning(f"Disconnected from AWS IoT Core: {rc}")
    
    def on_publish(self, client, userdata, mid):
        """MQTT publish callback."""
        logger.debug(f"Message published: {mid}")
    
    def create_devices(self) -> None:
        """Create simulated IoT devices."""
        for i in range(self.num_devices):
            device_id = f"device_{i:03d}"
            device_type = random.choice(self.device_types)
            
            device = IoTDevice(
                device_id=device_id,
                device_type=device_type,
                mqtt_client=self.mqtt_client
            )
            
            self.devices.append(device)
            logger.info(f"Created device {device_id} of type {device_type}")
    
    def start_benign_simulation(self) -> None:
        """Start benign traffic simulation for all devices."""
        logger.info("Starting benign traffic simulation...")
        
        for device in self.devices:
            device.start_benign_behavior()
        
        logger.info(f"Started benign simulation for {len(self.devices)} devices")
    
    def launch_attack_scenario(self, attack_type: str, 
                             num_compromised: int = None,
                             duration: int = 60) -> None:
        """Launch a specific attack scenario."""
        if num_compromised is None:
            num_compromised = random.randint(1, max(1, len(self.devices) // 3))
        
        # Select random devices to compromise
        compromised_devices = random.sample(self.devices, num_compromised)
        
        logger.warning(f"Launching {attack_type} attack on {num_compromised} devices for {duration} seconds")
        
        for device in compromised_devices:
            device.start_attack_behavior(attack_type, duration)
        
        return compromised_devices
    
    def run_continuous_simulation(self, attack_frequency: int = 300) -> None:
        """Run continuous simulation with periodic attacks."""
        logger.info("Starting continuous attack simulation...")
        self.running = True
        
        # Start benign simulation
        self.start_benign_simulation()
        
        # Periodic attack simulation
        while self.running:
            try:
                # Random interval between attacks
                sleep_time = random.uniform(attack_frequency * 0.5, attack_frequency * 1.5)
                logger.info(f"Waiting {sleep_time:.1f} seconds until next attack scenario...")
                time.sleep(sleep_time)
                
                if not self.running:
                    break
                
                # Random attack scenario
                attack_type = random.choice(self.attack_scenarios)
                num_compromised = random.randint(1, max(1, len(self.devices) // 4))
                duration = random.randint(30, 120)
                
                self.launch_attack_scenario(attack_type, num_compromised, duration)
                
            except KeyboardInterrupt:
                logger.info("Received interrupt signal, stopping simulation...")
                self.stop()
                break
            except Exception as e:
                logger.error(f"Error in continuous simulation: {e}")
                time.sleep(30)
    
    def stop(self) -> None:
        """Stop attack emulation."""
        logger.info("Stopping attack emulation...")
        self.running = False
        
        # Stop all devices
        for device in self.devices:
            device.stop()
        
        # Disconnect MQTT client
        if self.mqtt_client:
            self.mqtt_client.loop_stop()
            self.mqtt_client.disconnect()
        
        logger.info("Attack emulation stopped")


def main():
    """Main function to run attack emulation."""
    parser = argparse.ArgumentParser(description='IoT Attack Emulation for Threat Detection')
    parser.add_argument('--endpoint', required=True, help='AWS IoT Core endpoint')
    parser.add_argument('--devices', type=int, default=10, help='Number of simulated devices')
    parser.add_argument('--attack-frequency', type=int, default=300, 
                       help='Average seconds between attack scenarios')
    parser.add_argument('--mode', choices=['continuous', 'single'], default='continuous',
                       help='Simulation mode')
    parser.add_argument('--attack-type', choices=[
        'ddos_flood', 'data_exfiltration', 'c2_beacon', 
        'scanning', 'backdoor', 'injection'
    ], help='Single attack type (for single mode)')
    parser.add_argument('--duration', type=int, default=60, 
                       help='Attack duration in seconds (for single mode)')
    
    args = parser.parse_args()
    
    # Create attack emulator
    emulator = AttackEmulator(
        iot_endpoint=args.endpoint,
        num_devices=args.devices
    )
    
    try:
        # Setup MQTT connection
        emulator.setup_mqtt_client()
        time.sleep(2)  # Allow connection to establish
        
        # Create devices
        emulator.create_devices()
        
        if args.mode == 'continuous':
            # Run continuous simulation
            emulator.run_continuous_simulation(args.attack_frequency)
        else:
            # Run single attack scenario
            if not args.attack_type:
                args.attack_type = random.choice(emulator.attack_scenarios)
            
            emulator.start_benign_simulation()
            time.sleep(10)  # Allow benign traffic to establish
            
            logger.info(f"Running single attack scenario: {args.attack_type}")
            emulator.launch_attack_scenario(args.attack_type, duration=args.duration)
            
            # Keep running for attack duration + buffer
            time.sleep(args.duration + 30)
            
    except KeyboardInterrupt:
        logger.info("Received interrupt signal")
    except Exception as e:
        logger.error(f"Error in main execution: {e}")
        raise
    finally:
        emulator.stop()


if __name__ == "__main__":
    main()